### PR TITLE
Added checks to skip rollout status for eligible workloads

### PR DESCRIPTION
**Pull request by koushdey**
... if updateStrategy is not RollingUpdate

Fixes: https://github.com/microsoft/azure-pipelines-tasks/issues/12589

### CODE REVIEW

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts [hunk 1]
```diff
@@ -205,7 +205,15 @@ export function getResources(filePaths: string[], filterResourceTypes: string[])
         const fileContents = fs.readFileSync(filePath);
         yaml.safeLoadAll(fileContents, function (inputObject) {
             const inputObjectKind = inputObject ? inputObject.kind : '';
+            let inputObjectStrategyType = '';
+            if(inputObject && inputObject.spec && inputObject.spec.updateStrategy){
+                inputObjectStrategyType = inputObject.spec.updateStrategy.type;
+            }
             if (filterResourceTypes.filter(type => isEqual(inputObjectKind, type, StringComparer.OrdinalIgnoreCase)).length > 0) {
+                if (checkStrategy && workloadTypesWithRolloutStatus.indexOf(inputObjectKind.toLowerCase()) >= 0 && !(inputObjectStrategyType === '' || isEqual(inputObjectStrategyType, "RollingUpdate", StringComparer.OrdinalIgnoreCase))) {
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:213]: Can you make this condition a little readable?

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts [hunk 2]
```diff
@@ -194,7 +194,7 @@ export function updateSelectorLabels(inputObject: any, newLabels: Map<string, st
     setSpecSelectorLabels(inputObject, existingLabels);
 }
 
-export function getResources(filePaths: string[], filterResourceTypes: string[]): Resource[] {
+export function getResources(filePaths: string[], filterResourceTypes: string[], checkStrategy = false): Resource[] {
```
[the remark this answers is not in the document]
Comment (koushdey) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:197]: This function is later used to fetch ingress [resources](https://github.com/microsoft/azure-pipelines-tasks/blob/a9edb75f2f661c7a97d60a8c2dc9246172cc7bad/Tasks/KubernetesManifestV0/src/utils/DeploymentHelper.ts#L40) after the checkManifestStability(). Hence I used this variable to skip the check during ingress resource fetching as we need this check only for rollout status that is part of checkManifestStability().

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts [hunk 3]
```diff
@@ -205,7 +205,15 @@ export function getResources(filePaths: string[], filterResourceTypes: string[])
         const fileContents = fs.readFileSync(filePath);
         yaml.safeLoadAll(fileContents, function (inputObject) {
             const inputObjectKind = inputObject ? inputObject.kind : '';
+            let inputObjectStrategyType = '';
+            if(inputObject && inputObject.spec && inputObject.spec.updateStrategy){
```
[the remark this answers is not in the document]
Comment (koushdey) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:209]: ok.

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts (hunk 1 above)
Comment (koushdey) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:213]: As per chat I will make nested if conditions to make it readable.

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ds-ms) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:197]: Rename it to checkWorkloadStrategy

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts [hunk 4]
```diff
@@ -205,7 +205,20 @@ export function getResources(filePaths: string[], filterResourceTypes: string[])
         const fileContents = fs.readFileSync(filePath);
         yaml.safeLoadAll(fileContents, function (inputObject) {
             const inputObjectKind = inputObject ? inputObject.kind : '';
+            let inputObjectStrategyType = '';
+            if (inputObject && inputObject.spec && inputObject.spec.updateStrategy) {
+                inputObjectStrategyType = inputObject.spec.updateStrategy.type;
+            }
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:211]: Set else to `RollingUpdate` maybe?

Comment (ds-ms) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:211]: Add the check from line # 215 here. And remove it from there. And set the inputObjectStrategyType to `RollingUpdate` only if it's a workload. That should simplify the logic there.

File: Tasks/KubernetesManifestV0/src/utils/DeploymentHelper.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ export async function deploy(kubectl: Kubectl, manifestFilePaths: string[], depl
     const deployedManifestFiles = deployManifests(inputManifestFiles, kubectl, isCanaryDeploymentStrategy(deploymentStrategy));
 
     // check manifest stability
-    const resourceTypes: Resource[] = KubernetesObjectUtility.getResources(deployedManifestFiles, models.deploymentTypes.concat([KubernetesConstants.DiscoveryAndLoadBalancerResource.service]));
+    const resourceTypes: Resource[] = KubernetesObjectUtility.getResources(deployedManifestFiles, models.deploymentTypes.concat([KubernetesConstants.DiscoveryAndLoadBalancerResource.service]), true);
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/DeploymentHelper.ts:36]: Add a separate step for fetching `services`.

**Comment by ds-ms** [comment]
Move this logic to https://github.com/microsoft/azure-pipelines-tasks/blob/91ed11388bfa1462fd999a7849c3c8b5db08eccb/Tasks/Common/kubernetes-common-v2/kubernetesmanifestutility.ts#L9

Might be cleaner.

cc: @shigupt202

**Comment by ds-ms** [comment]
@koushdey could you add a service.yaml to the tests and validate the stability?

**Comment by shigupt202** [comment]
> Move this logic to
> 
> https://github.com/microsoft/azure-pipelines-tasks/blob/91ed11388bfa1462fd999a7849c3c8b5db08eccb/Tasks/Common/kubernetes-common-v2/kubernetesmanifestutility.ts#L9
> 
> Might be cleaner.
> 
> cc: @shigupt202

Agreed. There are other non-workloadTypesWithRolloutStatus for which we check manifest stability, which will be missed if you filter them out in getResources itself. Also, it's cleaner to put such checks outside of the getResource().

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts (hunk 4 above)
Comment (koushdey) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:211]: Done.

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (koushdey) [reply Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:197]: Done

**Comment by shigupt202** [comment]
LGTM. Make sure to bump the version of all the tasks using kubernetes-common-v2, to fix the failing checks.

File: Tasks/KubernetesManifestV0/Tests/L0.ts [hunk 6]
```diff
@@ -123,7 +123,7 @@ describe('Kubernetes Manifests Suite', function () {
         tr.run();
         assert(tr.succeeded, 'task should have succeeded');
         assert(tr.stdout.indexOf('nginx-deployment created') != -1, 'deployment is created');
-        assert(tr.stdout.indexOf('deployment "nginx-deployment" successfully rolled out') != -1, 'deployment is successfully rolled out');
+        assert(tr.stdout.indexOf('Rollout status has been skipped for Deployment as only updateStartegy:\'RollingUpdate\' is allowed') != -1, 'deployment rollout status skipped');
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/Tests/L0.ts:126]: Could you assert the `service` stability check as well?

File: Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts [hunk 7]
```diff
@@ -205,10 +205,24 @@ export function getResources(filePaths: string[], filterResourceTypes: string[])
...
+                if (!isEqual(inputObjectStrategyType, "RollingUpdate", StringComparer.OrdinalIgnoreCase)) {
+                    rollingUpdateExists = false;
+                }
+            }
             if (filterResourceTypes.filter(type => isEqual(inputObjectKind, type, StringComparer.OrdinalIgnoreCase)).length > 0) {
                 const resource = {
                     type: inputObject.kind,
-                    name: inputObject.metadata.name
+                    name: inputObject.metadata.name,
+                    isStrategyRollingUpdate: rollingUpdateExists
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/KubernetesObjectUtility.ts:225]: NIT: call this variable `isStrategyRollingUpdate` only?

**Pull request merged**